### PR TITLE
Set Zone Floor Area = Zone CalcFloorArea in SurfaceGeometr

### LINKED ISSUE #9515: Zero zone floor area causes E+ crash by running debug
Issue overview
--------------
The release runs the defect file fine. However, when run with the debugger zero floor area causes program crashed in Line 525 in the SurfaceGeometry module

                    ZFlrAvg += ((Z1 + Z2) / 2.0) * (thisSurface.Area / thisZone.FloorArea);

When debug the program, found zone calcFloorArea > 0 and FloorArea = 0. 
### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ X] Defect file added (list location of defect file here)
 - [ X] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by lgu1234**
Pull request overview
---------------------

Fixes #9515 
 When state->dataHeatBal->Zone(1).CalcFloorArea > 0 and  state->dataHeatBal->Zone(1).FloorArea = 0, the program crashed. The fix is to set the state->dataHeatBal->Zone(1).FloorArea = state->dataHeatBal->Zone(1).CalcFloorArea

The test files are enclosed.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [X ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ X] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by lgu1234** [comment]
Here are test files:
[9515.zip](https://github.com/NREL/EnergyPlus/files/9055534/9515.zip)

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 1]
```diff
@@ -2469,6 +2469,9 @@ namespace SurfaceGeometry {
                                                   "Unable to apportion the zone floor area. Space Floor Areas are zero.");
                             }
                         }
+                    } else {
+                        if (state.dataHeatBal->Zone(ZoneNum).CalcFloorArea > 0.0)
+                            state.dataHeatBal->Zone(ZoneNum).FloorArea = state.dataHeatBal->Zone(ZoneNum).CalcFloorArea;
```
Comment (lgu1234) [inline src/EnergyPlus/SurfaceGeometry.cc:2474]: Set Zone FloorArea = Zone CalcFlorrArea. Otherwise, the program is crashed with debug due to ZoneFloorArea = 0.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc [hunk 2]
```diff
@@ -8860,6 +8860,167 @@ TEST_F(EnergyPlusFixture, SurfaceGeometry_ZoneAndSpaceAreas)
...
+    state->dataSurfaceGeometry->SinZoneRelNorth(1) = std::sin(-state->dataHeatBal->Zone(1).RelNorth * DataGlobalConstants::DegToRadians);
+    state->dataSurfaceGeometry->CosBldgRelNorth = 1.0;
+    state->dataSurfaceGeometry->SinBldgRelNorth = 0.0;
+    GetSurfaceData(*state, ErrorsFound); // setup zone geometry and get zone data
+    EXPECT_FALSE(ErrorsFound);           // expect no errors
+
+    EXPECT_NEAR(state->dataHeatBal->Zone(1).FloorArea, 1.0, 0.001);
+    EXPECT_NEAR(state->dataHeatBal->Zone(1).CalcFloorArea, 1.0, 0.001);
+}
+
```
Comment (lgu1234) [inline tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:9023]: Unit test to show zone Floor Area are assigned correctly and is equal to the CalcFloorArea.

**Comment by Myoldmopar** [comment]
@lgu1234 can you explain what was happening before?  Your description explains the underlying condition (CalcFloorArea > 0), but what input conditions would have triggered this crash?  Or how was this working before?  Thanks!

**Comment by lgu1234** [comment]
Here is zipped file, containing test file.
[9515.zip](https://github.com/NREL/EnergyPlus/files/9103729/9515.zip)

I tried to upload EnergyPluDevSupport. For some reason, the defect files are not shown in GitHub.

The trigger input condition is that the zone floor area is 0;

  Zone,
    RA Duct Zone_1,          !- Name
    0,                       !- Direction of Relative North {deg}
    0,                       !- X Origin {m}
    0,                       !- Y Origin {m}
    0,                       !- Z Origin {m}
    ,                        !- Type
    ,                        !- Multiplier
    0,                       !- Ceiling Height {m}
    0,                       !- Volume {m3}
    0;                       !- Floor Area {m2}

Please let me know if you need more information.

**Comment by lgu1234** [comment]
@Myoldmopar I made a reply above.

**Comment by Myoldmopar** [comment]
> 0, !- Ceiling Height {m}
> 0, !- Volume {m3}
> 0; !- Floor Area {m2}

Interesting. I would have thought these inputs would be either blank or greater than zero.  Is there any thought to changing to inputs to force a non-blank value to be greater than zero?  What do you think?

**Comment by lgu1234** [comment]
@Myoldmopar Based on idd, 0 value is allowed:

  N8 , \field Volume
       \note If this field is 0.0, negative or autocalculate, then the volume of the zone
       \note is automatically calculated and used in subsequent calculations.
       \note If this field is positive, then the number entered here will be used.
       \units m3
       \type real
       \autocalculatable
       \default autocalculate

When zero value is entered, the floor area should be calculated based on zone CalcFloorArea.

**Comment by Myoldmopar** [comment]
So then what happens in develop right now?  If I enter zero on any input file, it will definitely crash?  I assumed this was just some corner case, not completely broken.

**Comment by lgu1234** [comment]
@Myoldmopar I don't know if this issue is applied to every zone input or not. I hope not. However, the test file show the crash and may be a corner case.

File: src/EnergyPlus/SurfaceGeometry.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/SurfaceGeometry.cc:2474]: I wonder why we need a separate CalcFloorArea member variable then.  I won't drag out this tiny fix, but I would have expected `FloorArea` to be assigned zero (or `DataSizing::AutoCalculate`), and then calculated at the right time if it needs to be.  Oh well, this is fine.

File: tst/EnergyPlus/unit/SurfaceGeometry.unit.cc (hunk 2 above)
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/SurfaceGeometry.unit.cc:9023]: So assuming we have two different variables holding floor area, then this unit test properly checks that.  It's OK for now.

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
I could poke at this further to sate my curiosity and figure out why we have a different variable there, but it's probably better use of time to just let it go.  It avoids the crash, so we're good.  Thanks @lgu1234, merging.

**Comment by lgu1234** [comment]
@Myoldmopar I tested an example file of SmallOffice_CentralDOAS.idf. If I change input filed values of Zone into zeros for Ceiling Height, Volume and Floor Area as below, the program still crashed. Therefore, my guess is a general one.

  Zone,
    Perimeter_ZN_4,          !- Name
    0.0000,                  !- Direction of Relative North {deg}
    0.0000,                  !- X Origin {m}
    0.0000,                  !- Y Origin {m}
    0.0000,                  !- Z Origin {m}
    1,                       !- Type
    1,                       !- Multiplier
    0.0,                        !- Ceiling Height {m}
    0.0,                        !- Volume {m3}
!    autocalculate,           !- Floor Area {m2}
    0.0,           !- Floor Area {m2}
    ,                        !- Zone Inside Convection Algorithm
    ,                        !- Zone Outside Convection Algorithm
    Yes;                     !- Part of Total Floor Area

**Comment by mjwitte** [comment]
@Myoldmopar Just to clear up some things.
1. It looks like I broke this by being a little too efficient when adding Space. Before Space, `Zone.FloorArea` was set to the calculated area (sum of floor surface areas) first, then the checks for user entered floor are were done and would change `Zone.FloorArea` if needed. When Space was added, `Zone.FloorArea` did not get set to the calculated area up front as before. This fix restores that assignment (in a different place, but it covers it.
2. The logic here *could* be made a bit simpler, but this is one-time per zone during setup.
3. You are correct that `Zone.CalcFloorArea` is only used locally inside this `for Zone` loop, so it could be made local instead of being a member of `Zone`.  `space.calcFloorArea` on the other hand, does need to be kept around from one loop to the next here.

**Comment by mjwitte** [comment]
For the record, this is the [code before Space](https://github.com/NREL/EnergyPlus/blob/bba177503b1df5dcc2fc09dc2ed72706efa528a7/src/EnergyPlus/SurfaceGeometry.cc#L2368) was added.
